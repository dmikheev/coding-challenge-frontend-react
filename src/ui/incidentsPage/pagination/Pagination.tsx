import { Button } from '@material-ui/core';
import React from 'react';
import styled from 'styled-components';

const VISIBLE_PAGE_NUM_COUNT = 5;

interface IPaginationProps {
  className?: string;
  activePage: number;
  itemsCountPerPage: number;
  totalItemsCount: number;
  onChange(page: number): void;
}
const Pagination: React.FC<IPaginationProps> = ({
  className,
  activePage,
  itemsCountPerPage,
  totalItemsCount,
  onChange,
}) => {
  const numButtonsHtml: React.ReactNode[] = [];
  const visiblePageCountToOneSide = Math.floor((VISIBLE_PAGE_NUM_COUNT - 1) / 2);
  const totalPagesCount = Math.ceil(totalItemsCount / itemsCountPerPage);
  const firstPageNum = Math.max(1, activePage - visiblePageCountToOneSide);
  const lastPageNum = Math.min(totalPagesCount, activePage + visiblePageCountToOneSide);
  for (let i = firstPageNum; i <= lastPageNum; i++) {
    numButtonsHtml.push((
      <ButtonWithNum
        key={i}
        variant="contained"
        color={i === activePage ? 'secondary' : undefined}
        onClick={() => onChange(i)}
      >
        {i}
      </ButtonWithNum>
    ));
  }

  return (
    <StyledWrap className={className}>
      <ButtonWithText disabled={activePage === 1} variant="contained" onClick={() => onChange(1)}>
        {'<< First'}
      </ButtonWithText>
      <ButtonWithText
        disabled={activePage === 1}
        variant="contained"
        onClick={() => onChange(activePage - 1)}
      >
        {'< Prev'}
      </ButtonWithText>
      {numButtonsHtml}
      <ButtonWithText
        disabled={activePage === totalPagesCount}
        variant="contained"
        onClick={() => onChange(activePage + 1)}
      >
        {'Next >'}
      </ButtonWithText>
      <ButtonWithText
        disabled={activePage === totalPagesCount}
        variant="contained"
        onClick={() => onChange(totalPagesCount)}
      >
        {'Last >>'}
      </ButtonWithText>
    </StyledWrap>
  );
};
export default Pagination;

const StyledWrap = styled.div`
  & > *:not(:last-child) {
    margin-right: 20px;
  }
`;

const ButtonWithText = styled(Button)`
  min-width: 100px;
`;

const ButtonWithNum = styled(Button)`
  min-width: auto;
`;
