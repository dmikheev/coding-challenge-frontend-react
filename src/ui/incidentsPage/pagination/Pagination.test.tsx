import { cleanup, fireEvent, render } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Pagination from './Pagination';

function nop () {}

afterEach(cleanup);

it('renders correctly', () => {
  const wrapper = shallow(
    <Pagination activePage={5} itemsCountPerPage={10} totalItemsCount={100} onChange={nop}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly when totalPages < 5', () => {
  const wrapper = shallow(
    <Pagination activePage={1} itemsCountPerPage={10} totalItemsCount={30} onChange={nop}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly when activePage = 1', () => {
  const wrapper = shallow(
    <Pagination activePage={1} itemsCountPerPage={10} totalItemsCount={100} onChange={nop}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly when activePage is the last page', () => {
  const wrapper = shallow(
    <Pagination activePage={10} itemsCountPerPage={10} totalItemsCount={100} onChange={nop}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('calls onChange with the right page number on a number button click', () => {
  const onChange = jest.fn();
  const { queryByText } = render(
    <Pagination activePage={5} itemsCountPerPage={10} totalItemsCount={100} onChange={onChange}/>
  );

  const firstButton = queryByText('7');
  expect(firstButton).toBeTruthy();
  fireEvent.click(firstButton!);

  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toHaveBeenLastCalledWith(7);
});

it('calls onChange with 1 on the "First" button click', () => {
  const onChange = jest.fn();
  const { queryByText } = render(
    <Pagination activePage={5} itemsCountPerPage={10} totalItemsCount={100} onChange={onChange}/>
  );

  const firstButton = queryByText('<< First');
  expect(firstButton).toBeTruthy();
  fireEvent.click(firstButton!);

  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toHaveBeenLastCalledWith(1);
});

it('calls onChange with prev page on the "Prev" button click', () => {
  const activePage = 5;
  const onChange = jest.fn();
  const { queryByText } = render(
    <Pagination activePage={activePage} itemsCountPerPage={10} totalItemsCount={100} onChange={onChange}/>
  );

  const firstButton = queryByText('< Prev');
  expect(firstButton).toBeTruthy();
  fireEvent.click(firstButton!);

  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toHaveBeenLastCalledWith(activePage - 1);
});

it('calls onChange with next page on the "Next" button click', () => {
  const activePage = 5;
  const onChange = jest.fn();
  const { queryByText } = render(
    <Pagination activePage={activePage} itemsCountPerPage={10} totalItemsCount={100} onChange={onChange}/>
  );

  const firstButton = queryByText('Next >');
  expect(firstButton).toBeTruthy();
  fireEvent.click(firstButton!);

  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toHaveBeenLastCalledWith(activePage + 1);
});

it('calls onChange with the last page on the "Last" button click', () => {
  const onChange = jest.fn();
  const { queryByText } = render(
    <Pagination activePage={5} itemsCountPerPage={10} totalItemsCount={100} onChange={onChange}/>
  );

  const firstButton = queryByText('Last >>');
  expect(firstButton).toBeTruthy();
  fireEvent.click(firstButton!);

  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toHaveBeenLastCalledWith(10);
});
