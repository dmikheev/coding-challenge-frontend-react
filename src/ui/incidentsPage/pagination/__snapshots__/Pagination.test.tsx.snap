// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    &lt;&lt; First
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    &lt; Prev
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="3"
    onClick={[Function]}
    variant="contained"
  >
    3
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="4"
    onClick={[Function]}
    variant="contained"
  >
    4
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    color="secondary"
    key="5"
    onClick={[Function]}
    variant="contained"
  >
    5
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="6"
    onClick={[Function]}
    variant="contained"
  >
    6
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="7"
    onClick={[Function]}
    variant="contained"
  >
    7
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    Next &gt;
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    Last &gt;&gt;
  </Styled(WithStyles(ForwardRef(Button)))>
</styled.div>
`;

exports[`renders correctly when activePage = 1 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    &lt;&lt; First
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    &lt; Prev
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    color="secondary"
    key="1"
    onClick={[Function]}
    variant="contained"
  >
    1
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="2"
    onClick={[Function]}
    variant="contained"
  >
    2
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="3"
    onClick={[Function]}
    variant="contained"
  >
    3
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    Next &gt;
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    Last &gt;&gt;
  </Styled(WithStyles(ForwardRef(Button)))>
</styled.div>
`;

exports[`renders correctly when activePage is the last page 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    &lt;&lt; First
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    &lt; Prev
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="8"
    onClick={[Function]}
    variant="contained"
  >
    8
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="9"
    onClick={[Function]}
    variant="contained"
  >
    9
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    color="secondary"
    key="10"
    onClick={[Function]}
    variant="contained"
  >
    10
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    Next &gt;
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    Last &gt;&gt;
  </Styled(WithStyles(ForwardRef(Button)))>
</styled.div>
`;

exports[`renders correctly when totalPages < 5 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    &lt;&lt; First
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    &lt; Prev
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    color="secondary"
    key="1"
    onClick={[Function]}
    variant="contained"
  >
    1
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="2"
    onClick={[Function]}
    variant="contained"
  >
    2
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    key="3"
    onClick={[Function]}
    variant="contained"
  >
    3
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    Next &gt;
  </Styled(WithStyles(ForwardRef(Button)))>
  <Styled(WithStyles(ForwardRef(Button)))
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    Last &gt;&gt;
  </Styled(WithStyles(ForwardRef(Button)))>
</styled.div>
`;
